### PR TITLE
add nonuniform dro

### PR DESCRIPTION
**Pull request by guyichen09**

### CODE REVIEW

File: src/plugins/risk_measures.jl [hunk 1]
```diff
@@ -329,15 +329,109 @@ uniform.
...
+        return 0.0
+    end
+    if !is_minimization
+        z = -z
+    end
+    # step 1
+    K = collect(1:m)
+    # check if nomial probability is 0
+    for i in K
+        if q[i] ≈ 0
```
Comment (odow) [inline src/plugins/risk_measures.jl:349]: Maybe we just want `if isapprox(q[I], 0.0, atol = 1e-10)`.

File: src/plugins/risk_measures.jl [hunk 2]
```diff
@@ -329,15 +329,109 @@ uniform.
...
+    end
+    #update m
+    m = length(K)
+    # use this to store the index popped out of K
+    not_in_K = []
+    # step 2
+    while length(K) > 1
+        # step 2(a)
+        z_bar = sum(z[i] for i in K) / length(K)
+        #println(z_bar)
```
Comment (odow) [inline src/plugins/risk_measures.jl:362]: Please remove the unneeded comments

File: test/plugins/risk_measures.jl [hunk 3]
```diff
@@ -190,13 +190,14 @@ end
     @testset "Non-uniform distribution" begin
         risk_adjusted_probability = Vector{Float64}(undef, 5)
         @test_throws Exception SDDP.adjust_probability(
```
Comment (odow) [inline test/plugins/risk_measures.jl:192]: Remove `@test_throws Exception `.

File: test/plugins/risk_measures.jl [hunk 4]
```diff
@@ -190,13 +190,14 @@ end
...
-            SDDP.ModifiedChiSquared(0.1),
+            SDDP.ModifiedChiSquared(0.0),
             risk_adjusted_probability,
             [0.1, 0.2, 0.3, 0.2, 0.2],
             [:a, :b, :c, :d, :e],
             [-2.0, -1.0, -3.0, -4.0, -5.0],
             true,
         )
+        @test risk_adjusted_probability ≈ [0.2, 0.2, 0.2, 0.2, 0.2] atol = 1e-6
```
Comment (odow) [inline test/plugins/risk_measures.jl:200]: This should test against the nominal probability.

File: src/plugins/risk_measures.jl [hunk 5]
```diff
@@ -329,15 +329,109 @@ uniform.
...
+        s² = sum(z[i]^2 - z_bar^2 for i in K) / length(K)
+        #println(s²)
+        # step 2(b)
+        if length(K) == m
+            for i in K
+                p[i] = q[i] +
+                (z[i] - z_bar) / (sqrt(m) * sqrt(s²)) * measure.r
```
Comment (odow) [inline src/plugins/risk_measures.jl:369]: Small issue, please put this all on one line, or indent the second line

File: src/plugins/risk_measures.jl [hunk 6]
```diff
@@ -329,15 +329,109 @@ uniform.
...
+    for i in K
+        if q[i] ≈ 0
+            p[i] = 0
+            splice!(K, i)
+        end
+    end
+    #update m
+    m = length(K)
+    # use this to store the index popped out of K
+    not_in_K = []
```
Comment (odow) [inline src/plugins/risk_measures.jl:357]: Please type this to `not_in_K = Int[]`.

File: src/plugins/risk_measures.jl [hunk 7]
```diff
@@ -329,15 +329,109 @@ uniform.
...
+        sum_qj_squared = 0
+        if length(not_in_K) > 0
+            sum_qj = sum(q[i] for i in not_in_K)
+            sum_qj_squared = sum(q[i]^2 for i in not_in_K)
+        end
+        len_k = length(K)
+        for i in negative_p
+            r² = (((-q[i] * len_k  - sum_qj)/((z[i] - z_bar)/sqrt(s²)))^2 + sum_qj_squared^2)/len_k + sum_qj_squared
+            append!(computed_r, r²)
+        end
```
Comment (odow) [inline src/plugins/risk_measures.jl:418]:
```Julia
computed_r = [
    (((-q[i] * len_k  - sum_qj)/((z[i] - z_bar)/sqrt(s²)))^2 + sum_qj_squared^2)/len_k + sum_qj_squared
    for i in negative_p
]
```

File: src/plugins/risk_measures.jl [hunk 8]
```diff
@@ -329,15 +329,109 @@ uniform.
...
+        #@show p
+        #print(computed_r)
+        i_K = negative_p[argmin(computed_r)]
+        append!(not_in_K, i_K)
+        filter!(e->e≠i_K,K)
```
Comment (odow) [inline src/plugins/risk_measures.jl:423]: `filter!(e -> e != i_K, K)`

File: src/plugins/risk_measures.jl [hunk 9]
```diff
@@ -329,15 +329,109 @@ uniform.
...
+    #update m
+    m = length(K)
+    # use this to store the index popped out of K
+    not_in_K = []
+    # step 2
+    while length(K) > 1
+        # step 2(a)
+        z_bar = sum(z[i] for i in K) / length(K)
+        #println(z_bar)
+        s² = sum(z[i]^2 - z_bar^2 for i in K) / length(K)
```
Comment (odow) [inline src/plugins/risk_measures.jl:363]: So it looks like we never use `s^2` directly? Only ever `sqrt(s^2)`? If so, maybe just add the sort here. The check for a too small std should avoid `sqrt(0)`.

File: test/plugins/risk_measures.jl (hunk 3 above)
Comment (guyichen09) [reply test/plugins/risk_measures.jl]: fixed

File: test/plugins/risk_measures.jl (hunk 4 above)
Comment (guyichen09) [reply test/plugins/risk_measures.jl]: fixed

File: src/plugins/risk_measures.jl (hunk 2 above)
Comment (guyichen09) [reply src/plugins/risk_measures.jl:362]: fixed

File: src/plugins/risk_measures.jl (hunk 1 above)
Comment (guyichen09) [reply src/plugins/risk_measures.jl:349]: fixed

File: src/plugins/risk_measures.jl (hunk 6 above)
Comment (guyichen09) [reply src/plugins/risk_measures.jl:357]: fixed

Comment (odow) [reply src/plugins/risk_measures.jl:357]: Easiest thing is to just push the commits first, then go through clicking "Resolve." No need to comment each one :)

File: test/plugins/risk_measures.jl [hunk 10]
```diff
@@ -199,6 +199,70 @@ end
...
+        SDDP.adjust_probability(
+            SDDP.ModifiedChiSquared(0.45),
+            risk_adjusted_probability,
+            [0.1, 0.2, 0.3, 0.3, 0.1],
+            [:a, :b, :c, :d, :e],
+            [-2.0, -1.0, -3.0, -4.0, -5.0],
+            false,
+        )
+        @test risk_adjusted_probability ≈ [0.0, 0.0, 0.323223, 0.676777, 0.0] atol = 1e-6
+
```
Comment (odow) [inline test/plugins/risk_measures.jl:265]: Looks like you are missing an `end` here

**Pull request merged**